uaclient/api/u/pro/services/list/v1.py

```diff
@@ -0,0 +1,201 @@
+# Services List
+import logging
+from typing import Dict, List, Optional
+
+from uaclient import exceptions, util
+from uaclient.api.api import APIEndpoint
+from uaclient.api.data_types import AdditionalInfo
+from uaclient.api.u.pro.status.is_attached.v1 import _is_attached
+from uaclient.config import UAConfig
+from uaclient.contract import get_available_resources
+from uaclient.data_types import (
+    BoolDataValue,
+    DataObject,
+    Field,
+    StringDataValue,
+    data_list,
+)
+from uaclient.entitlements import entitlement_factory
+from uaclient.entitlements.base import UAEntitlement
+from uaclient.entitlements.entitlement_status import ContractStatus
+from uaclient.files import machine_token
+
+LOG = logging.getLogger(util.replace_top_level_logger_name(__name__))
+
+
+class ServiceInfo(DataObject, AdditionalInfo):
+    fields = [
+        Field("name", StringDataValue, doc="Name of the service"),
+        Field(
+            "description", StringDataValue, doc="Description of the service"
+        ),
+        Field(
+            "available",
+            BoolDataValue,
+            doc="Service availability for a subscription",
+        ),
+        Field(
+            "entitled",
+            BoolDataValue,
+            doc="Entitlement of a service for a subscription",
+        ),
+    ]
+
+    def __init__(
+        self,
+        *,
+        name: str,
+        desc: str,
+        available: bool,
+        entitled: Optional[bool] = None
+    ):
+        self.name = name
+        self.description = desc
+        self.available = available
+        self.entitled = entitled
+
+
+class ServiceListResult(DataObject, AdditionalInfo):
+    fields = [
+        Field("services", data_list(ServiceInfo), doc="List of services"),
+    ]
+
+    def __init__(self, *, services: List[ServiceInfo]):
+        self.services = services
+
+
+def get_service_list() -> ServiceListResult:
+    return _get_service_list(UAConfig())
+
+
+def _get_service_list(cfg: UAConfig) -> ServiceListResult:
+    if not _is_attached(cfg).is_attached:
+        return unattached_services(cfg)
+    return attached_services(cfg)
+
+
+def attached_services(cfg: UAConfig) -> ServiceListResult:
+    services_list = []
+    machine_token_file = machine_token.get_machine_token_file(cfg)
+    resources = machine_token_file.machine_token.get("availableResources")
+    if not resources:
+        resources = get_available_resources(cfg)
+    inapplicable_resources = {
+        resource["name"]: resource.get("description")
+        for resource in sorted(resources, key=lambda x: x.get("name", ""))
+        if not resource.get("available")
+    }
+    for resource in resources:
+        try:
+            ent = entitlement_factory(cfg=cfg, name=resource.get("name", ""))
+        except exceptions.EntitlementNotFoundError:
+            continue
```

Comment:
shouldn't we have the same log message here for an unknown service ?

Reply:
Yes, updated, the message should be there for both attached and unattached